src/hotspot/share/gc/shenandoah/heuristics/shenandoahHeuristics.cpp

```diff
@@ -304,6 +436,155 @@ double ShenandoahHeuristics::time_since_last_gc() const {
 
 bool ShenandoahHeuristics::in_generation(ShenandoahHeapRegion* region) {
   return (_generation->generation_mode() == GLOBAL)
-      || (_generation->generation_mode() == YOUNG && region->affiliation() == YOUNG_GENERATION)
-      || (_generation->generation_mode() == OLD && region->affiliation() == OLD_GENERATION);
+    || (_generation->generation_mode() == YOUNG && region->affiliation() == YOUNG_GENERATION)
+    || (_generation->generation_mode() == OLD && region->affiliation() == OLD_GENERATION);
+}
+
+void ShenandoahHeuristics::prepare_for_old_collections() {
+  assert(_generation->generation_mode() == OLD, "This service only available for old-gc heuristics");
+  ShenandoahHeap* heap = ShenandoahHeap::heap();
+  uint free_regions = 0;
+  size_t cand_idx = 0;
+  size_t total_garbage = 0;
+  size_t num_regions = heap->num_regions();
+
+  RegionData* candidates = _region_data;
+  for (size_t i = 0; i < num_regions; i++) {
+    ShenandoahHeapRegion* region = heap->get_region(i);
+    if (!in_generation(region))
+      continue;
+    else {
+      size_t garbage = region->garbage();
+      total_garbage += garbage;
+
+      candidates[cand_idx]._region = region;
+      candidates[cand_idx]._garbage = garbage;
+      cand_idx++;
+    }
+  }
+
+  // Give special treatment to humongous regions.  Assume humongous regions is entirely
+  // garbage or entirely non-garbage.  Assume that a head humongous region and the associated
+  // humongous continuous regions are uniformly entirely garbage or entirely non-garbage.
+  //
+  // Sift garbage humongous regions to front, non-garbage humongous regions to end of array.
+  size_t first_non_humongous_empty = 0;
+  size_t first_humongous_non_empty = cand_idx;
+
+  // This loop is written as while rather than for because of
+  // suspected gcc error in translating/optimizing for-loop
```

Comment:
WTF, really? Do we have any indication of this?

Reply:
We will remove this comment. We saw some pretty strange behavior while debugging, but it turned out to be a bug in `rr`.

Reply:
Comment will be gone in next commit.

Reply:
Comment has been removed.